### Day 8: Handheld Halting

<body>
<p>
Your flight to the major airline hub reaches cruising altitude without incident. While you consider checking the in-flight menu for one of those drinks that come with a little umbrella, you are interrupted by the kid sitting next to you.

Their handheld game console won't turn on! They ask if you can take a look.

You narrow the problem down to a strange infinite loop in the boot code (your puzzle input) of the device. You should be able to fix it, but first you need to be able to run the code in isolation.

The boot code is represented as a text file with one instruction per line of text. Each instruction consists of an operation (acc, jmp, or nop) and an argument (a signed number like +4 or -20).
</p>


<li>
<ol>acc increases or decreases a single global value called the accumulator by the value given in the argument. For example, acc +7 would increase the accumulator by 7. The accumulator starts at 0. After an acc instruction, the instruction immediately below it is executed next.</ol>
<ol>jmp jumps to a new instruction relative to itself. The next instruction to execute is found using the argument as an offset from the jmp instruction; for example, jmp +2 would skip the next instruction, jmp +1 would continue to the instruction immediately below it, and jmp -20 would cause the instruction 20 lines above to be executed next.</ol>
<ol>nop stands for No OPeration - it does nothing. The instruction immediately below it is executed next.</ol>
</li>
</body>

In [14]:
import numpy as np
import copy

In [15]:
f = open("input08.txt")
inputs=f.read()
inputs = inputs.split("\n")
inputs = [s.split(" ") for s in inputs]

#### Star 1

Run your copy of the boot code. Immediately before any instruction is executed a second time <em>i.e. before infinite loop</em>, what value is in the accumulator?

In [16]:

def star1(inputs) :
    i = 0
    acc = 0
    visited = np.zeros(len(inputs), dtype=bool)
    while i < len(inputs):
        if(visited[i] != False):
            break
        visited[i] = True
        if inputs[i][0] == 'jmp':
            i += int(inputs[i][1])
        elif inputs[i][0] == 'acc':
            acc += int(inputs[i][1])
            i+=1
        else:
            i +=1
    return i, acc

        

In [17]:
star1(inputs)[1]

1475

<font color="Red">Your puzzle answer was 1475.</font>

#### Star 2

After some careful analysis, you believe that exactly one instruction is corrupted.

Somewhere in the program, either a jmp is supposed to be a nop, or a nop is supposed to be a jmp. (No acc instructions were harmed in the corruption of this boot code.)

The program is supposed to terminate by attempting to execute an instruction immediately after the last instruction in the file. By changing exactly one jmp or nop, you can repair the boot code and make it terminate correctly.

Fix the program so that it terminates normally by changing exactly one jmp (to nop) or nop (to jmp). What is the value of the accumulator after the program terminates?

In [18]:
#star2

swapped = np.zeros(len(inputs), dtype = (bool))
swapped
x =0
for i in range(len(inputs)):
    inputsA = copy.deepcopy(inputs)
    if not swapped[i]:
        swapped[i] =True
        if inputsA[i][0] == 'jmp':
            inputsA[i][0] = 'nop'
        elif inputsA[i][0] == 'nop':
            inputsA[i][0] ='jmp'
    x, acc = star1(inputsA)
    if(x == len(inputs)):
        break

print(acc)

1270


<font color="Red"> Your puzzle answer was 1270</font>